File: backend/models/artist_language_metadata.py
"""Models for artist language metadata review endpoints."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

LanguageClassification = Literal["single_language", "multilingual", "instrumental"]
LanguageOrigin = Literal["manual", "curated_seed", "legacy_import"]
EvidenceKind = Literal[
    "artist_profile",
    "artist_repertoire",
    "editorial_source",
    "track_credit",
    "track_language",
]
PerformerAttribution = Literal[
    "artist_vocal_confirmed",
    "artist_instrumental_confirmed",
    "track_language_only",
    "not_applicable",
]
SourceStatus = Literal["suggested", "approved", "rejected", "superseded"]
ReviewStatus = Literal["open", "approved", "rejected", "insufficient_evidence"]
ReviewAction = Literal["approve", "reject", "insufficient_evidence"]
LanguageBucketClassification = Literal[
    "single_language",
    "multilingual",
    "instrumental",
    "unknown",
]


class ArtistLanguageEvidenceInput(BaseModel):
    model_config = ConfigDict(extra="forbid")

    local_track_id: int | None = None
    claimed_language_code: str | None = None
    claimed_language_variant: str | None = None
    evidence_kind: EvidenceKind
    performer_attribution: PerformerAttribution
    evidence_url: str
    evidence_title: str
    evidence_summary: str


class ArtistLanguageSourceInput(BaseModel):
    model_config = ConfigDict(extra="forbid")

    classification: LanguageClassification
    primary_language_code: str | None = None
    language_variant: str | None = None
    raw_language: str | None = None
    evidence: list[ArtistLanguageEvidenceInput] = Field(default_factory=list)


class ArtistLanguageReviewCreateRequest(BaseModel):
    artist_id: int
    reason: str = "manual_research"


class ArtistLanguageReviewDecisionRequest(BaseModel):
    action: ReviewAction
    resolution_note: str


class ArtistLanguageBucket(BaseModel):
    key: str
    label: str
    classification: LanguageBucketClassification
    hours: float
    share_pct: float
    artist_count: int


class ArtistLanguageMissingItem(BaseModel):
    artist_id: int
    artist_name: str
    hours: float


class ArtistLanguageCoverageResponse(BaseModel):
    eligible_hours: float
    excluded_unattributed_hours: float
    classified_hours: float
    unknown_hours: float
    classified_pct: float
    unknown_pct: float
    buckets: list[ArtistLanguageBucket] = Field(default_factory=list)
    source_hours: dict[str, float] = Field(default_factory=dict)
    top_missing: list[ArtistLanguageMissingItem] = Field(default_factory=list)
    caveat: str


class ArtistLanguageEvidenceItem(BaseModel):
    evidence_id: int
    source_id: int
    local_track_id: int | None = None
    claimed_language_code: str | None = None
    claimed_language_variant: str | None = None
    evidence_kind: EvidenceKind
    performer_attribution: PerformerAttribution
    evidence_url: str
    evidence_title: str
    evidence_accessed_at: str
    evidence_summary: str
    created_at: str


class ArtistLanguageSourceItem(BaseModel):
    source_id: int
    artist_id: int
    classification: LanguageClassification
    primary_language_code: str | None = None
    language_variant: str | None = None
    raw_language: str | None = None
    origin: LanguageOrigin
    source_key: str
    status: SourceStatus
    replaces_source_id: int | None = None
    created_at: str
    updated_at: str
    evidence: list[ArtistLanguageEvidenceItem] = Field(default_factory=list)


class ArtistLanguageReviewItem(BaseModel):
    review_id: int
    artist_id: int
    artist_name: str
    suggested_source_id: int | None = None
    play_hours_snapshot: float
    reason: str
    status: ReviewStatus
    resolution_note: str | None = None
    reviewed_by: str | None = None
    reviewed_at: str | None = None
    created_at: str
    updated_at: str
    source: ArtistLanguageSourceItem | None = None


class ArtistLanguageReviewListResponse(BaseModel):
    items: list[ArtistLanguageReviewItem] = Field(default_factory=list)
    total: int = Field(default=0, ge=0)


class ArtistLanguageReviewMutationResponse(BaseModel):
    review_id: int
    review_status: ReviewStatus
    source_id: int | None = None
    source_status: SourceStatus | None = None

File: frontend/src/types/artist-language-metadata.ts
export type LanguageClassification = 'single_language' | 'multilingual' | 'instrumental'
export type LanguageOrigin = 'manual' | 'curated_seed' | 'legacy_import'
export type ArtistLanguageEvidenceKind =
  | 'artist_profile'
  | 'artist_repertoire'
  | 'editorial_source'
  | 'track_credit'
  | 'track_language'
export type ArtistLanguagePerformerAttribution =
  | 'artist_vocal_confirmed'
  | 'artist_instrumental_confirmed'
  | 'track_language_only'
  | 'not_applicable'
export type ArtistLanguageSourceStatus = 'suggested' | 'approved' | 'rejected' | 'superseded'
export type ArtistLanguageReviewStatus = 'open' | 'approved' | 'rejected' | 'insufficient_evidence'
export type ArtistLanguageReviewAction = 'approve' | 'reject' | 'insufficient_evidence'
export type LanguageBucketClassification =
  | LanguageClassification
  | 'unknown'

export interface ArtistLanguagePlayFilters {
  min_ms: number
  music_only: boolean
  merge_enabled: boolean
  dynamic_threshold: boolean
  max_merge_gap_minutes?: number
}

export interface ArtistLanguageEvidenceInput {
  local_track_id?: number | null
  claimed_language_code?: string | null
  claimed_language_variant?: string | null
  evidence_kind: ArtistLanguageEvidenceKind
  performer_attribution: ArtistLanguagePerformerAttribution
  evidence_url: string
  evidence_title: string
  evidence_summary: string
}

export interface ArtistLanguageSourceInput {
  classification: LanguageClassification
  primary_language_code?: string | null
  language_variant?: string | null
  raw_language?: string | null
  evidence?: ArtistLanguageEvidenceInput[]
}

export interface ArtistLanguageReviewCreateRequest {
  artist_id: number
  reason?: string
}

export interface ArtistLanguageReviewDecisionRequest {
  action: ArtistLanguageReviewAction
  resolution_note: string
}

export interface LanguageBucket {
  key: string
  label: string
  classification: LanguageBucketClassification
  hours: number
  share_pct: number
  artist_count: number
}

export interface ArtistLanguageMissingItem {
  artist_id: number
  artist_name: string
  hours: number
}

export interface ArtistLanguageCoverage {
  eligible_hours: number
  excluded_unattributed_hours: number
  classified_hours: number
  unknown_hours: number
  classified_pct: number
  unknown_pct: number
  buckets: LanguageBucket[]
  source_hours: Record<string, number>
  top_missing: ArtistLanguageMissingItem[]
  caveat: string
}

export type ArtistLanguageCoverageResponse = ArtistLanguageCoverage

export interface ArtistLanguageEvidenceItem {
  evidence_id: number
  source_id: number
  local_track_id: number | null
  claimed_language_code: string | null
  claimed_language_variant: string | null
  evidence_kind: ArtistLanguageEvidenceKind
  performer_attribution: ArtistLanguagePerformerAttribution
  evidence_url: string
  evidence_title: string
  evidence_accessed_at: string
  evidence_summary: string
  created_at: string
}

export interface ArtistLanguageSourceItem {
  source_id: number
  artist_id: number
  classification: LanguageClassification
  primary_language_code: string | null
  language_variant: string | null
  raw_language: string | null
  origin: LanguageOrigin
  source_key: string
  status: ArtistLanguageSourceStatus
  replaces_source_id: number | null
  created_at: string
  updated_at: string
  evidence: ArtistLanguageEvidenceItem[]
}

export interface ArtistLanguageReviewItem {
  review_id: number
  artist_id: number
  artist_name: string
  suggested_source_id: number | null
  play_hours_snapshot: number
  reason: string
  status: ArtistLanguageReviewStatus
  pre_review_recommendation?: string | null
  pre_review_confidence?: number | null
  pre_review_note?: string | null
  pre_reviewed_by?: string | null
  pre_reviewed_at?: string | null
  resolution_note: string | null
  reviewed_by: string | null
  reviewed_at: string | null
  created_at: string
  updated_at: string
  source: ArtistLanguageSourceItem | null
}

export interface ArtistLanguageReviewListResponse {
  items: ArtistLanguageReviewItem[]
  total: number
}

export interface ArtistLanguageReviewMutationResponse {
  review_id: number
  review_status: ArtistLanguageReviewStatus
  source_id: number | null
  source_status: ArtistLanguageSourceStatus | null
}
